plutus-core/src/Language/PlutusCore/Evaluation/Machine/Ck.hs

```diff
@@ -132,7 +145,14 @@ to look up a free variable in an environment: there's no CkValue for
 Var, so we can't report which variable caused the error.
 -}
 
-instance ToExMemory term => SpendBudget (CkCarryingM term uni fun) fun () term where
+instance MonadEmitter (CkCarryingM term uni fun s) where
+    emit str = do
+        mayLogsRef <- asks ckEnvMayEmitRef
+        case mayLogsRef of
+            Nothing      -> pure ()
+            Just logsRef -> lift . lift $ modifySTRef logsRef (`DList.snoc` str)
```

Comment:
I think I'm going to remove dispatching on `mayLogsRef` and instead store a function in an environment as that should be faster. But the current version is cleaner, I guess.

Reply:
> That answers the question I raised in the CEK machine, so thanks for anticipating me!

You mean storing `spendBudget` as a function instead of doing dispatching on the budgeting mode all the time? Here I talk about `emit`. I think we should do both. But then it probably makes to do both in the same PR and leave the current PR as is.

Reply:
> You mean storing `spendBudget` as a function instead of doing dispatching on the budgeting mode all the time? Here I talk about `emit`. 

It was `emit` that I was asking about (https://github.com/input-output-hk/plutus/pull/2757#discussion_r580845922).  I agree that there's no great need to change it in this PR, especially if the budgeting changes will be happening soon.